src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java

```diff
@@ -0,0 +1,879 @@
+import org.openstreetmap.atlas.geography.atlas.builder.RelationBean;
+import org.openstreetmap.atlas.geography.atlas.items.ItemType;
+import org.openstreetmap.atlas.geography.atlas.items.Line;
+import org.openstreetmap.atlas.geography.atlas.items.Point;
+import org.openstreetmap.atlas.geography.atlas.packed.PackedAtlasBuilder;
+import org.openstreetmap.atlas.geography.atlas.pbf.AtlasLoadingOption;
+import org.openstreetmap.atlas.geography.atlas.pbf.slicing.identifier.WaySectionIdentifierFactory;
+import org.openstreetmap.atlas.geography.atlas.pbf.store.PbfOneWay;
+import org.openstreetmap.atlas.geography.atlas.raw.slicing.temporary.TemporaryEdge;
+import org.openstreetmap.atlas.geography.atlas.raw.slicing.temporary.TemporaryNode;
+import org.openstreetmap.atlas.geography.sharding.Shard;
+import org.openstreetmap.atlas.geography.sharding.Sharding;
+import org.openstreetmap.atlas.tags.AtlasTag;
+import org.openstreetmap.atlas.tags.HighwayTag;
+import org.openstreetmap.atlas.tags.ManMadeTag;
+import org.openstreetmap.atlas.tags.RailwayTag;
+import org.openstreetmap.atlas.tags.RouteTag;
+import org.openstreetmap.atlas.tags.annotations.validation.Validators;
+import org.openstreetmap.atlas.utilities.collections.Iterables;
+import org.openstreetmap.atlas.utilities.time.Time;
+import org.openstreetmap.osmosis.core.domain.v0_6.Node;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * TODO
+ *
+ * @author mgostintsev
+ */
+public class WaySectionProcessor
+{
+    private static final Logger logger = LoggerFactory.getLogger(WaySectionProcessor.class);
+
+    private static final int MINIMUM_SHAPE_POINTS_TO_QUALIFY_AS_AREA = 3;
+    private static final int MINIMUM_POINTS_TO_QUALIFY_AS_A_LINE = 2;
+
+    private final Atlas rawAtlas;
+    private final Sharding shardingTree;
+    private final Function<Shard, Optional<Atlas>> atlasFetcher;
+    private final AtlasLoadingOption loadingOption;
+
+    /**
+     * Default constructor. Will section given raw {@link Atlas} file.
+     *
+     * @param rawAtlas
+     *            The raw {@link Atlas} to section
+     * @param loadingOption
+     *            The {@link AtlasLoadingOption} to use
+     */
+    public WaySectionProcessor(final Atlas rawAtlas, final AtlasLoadingOption loadingOption)
+    {
+        this(rawAtlas, loadingOption, null, null);
+    }
+
+    /**
+     * Sections the given raw {@link Atlas} and guarantees consistent identifiers for all Atlas
+     * files obtained by using the given fetcher policy. If the sharding tree and atlas fetcher
+     * function is not provided, then no expansion will be done and only the given atlas will be
+     * way-sectioned. Note: the sharding tree must the same as the one used to generate the input
+     * raw Atlas.
+     *
+     * @param rawAtlas
+     *            The raw {@link Atlas} to section
+     * @param loadingOption
+     *            The {@link AtlasLoadingOption} to use
+     * @param shardingTree
```

Comment:
Naming nit here. A sharding object does not have to be a tree. The most common one is the quad tree, but for example the `SlippyTileSharding` is just all tiles at the same zoom level. I would call it just `sharding`!

Reply:
Good point!